engine/src/em-event.js

```diff
@@ -27,57 +27,76 @@ mergeInto(LibraryManager.library, {
 
+		// text element used to capture input events
+		_inputelement: null,
+
 		// This function is used to call a function for each event
 		// type to handler mapping defined for LiveCode on Emscripten.
 		//
-		// The <func> must take two arguments: the event type, and the
-		// handler function used to handle that event type.
+		// The <func> must take three arguments: the event type, the
+		// handler function used to handle that event type, and a
+		// boolean indicating whether the event handlershould be
+		// attached to the hidden input control.
 		_eventForEach: function(func) {
 
 			// Master mapping from event types to handler functions.
 			var mapping = [
-				['focus', LiveCodeEvents._handleFocusEvent],
-				['blur', LiveCodeEvents._handleFocusEvent],
+				['focus', LiveCodeEvents._handleFocusEvent, false],
+				['blur', LiveCodeEvents._handleFocusEvent, false],
 
-				['mousemove', LiveCodeEvents._handleMouseEvent],
-				['mousedown', LiveCodeEvents._handleMouseEvent],
-				['mouseup', LiveCodeEvents._handleMouseEvent],
-				['mouseenter', LiveCodeEvents._handleMouseEvent],
-				['mouseleave', LiveCodeEvents._handleMouseEvent],
+				['mousemove', LiveCodeEvents._handleMouseEvent, false],
+				['mousedown', LiveCodeEvents._handleMouseEvent, false],
+				['mouseup', LiveCodeEvents._handleMouseEvent, false],
+				['mouseenter', LiveCodeEvents._handleMouseEvent, false],
+				['mouseleave', LiveCodeEvents._handleMouseEvent, false],
 
 				// FIXME "keypress" events are deprecated
-				['keypress', LiveCodeEvents._handleKeyboardEvent],
-				['keyup', LiveCodeEvents._handleKeyboardEvent],
-				['keydown', LiveCodeEvents._handleKeyboardEvent],
+				['keypress', LiveCodeEvents._handleKeyboardEvent, true],
+				['keyup', LiveCodeEvents._handleKeyboardEvent, true],
+				['keydown', LiveCodeEvents._handleKeyboardEvent, true],
 
-				['input', LiveCodeEvents._handleInput],
-				['beforeinput', LiveCodeEvents._handleInput],
+				['input', LiveCodeEvents._handleInput, true],
+				['beforeinput', LiveCodeEvents._handleInput, true],
 
-				['compositionstart', LiveCodeEvents._handleComposition],
-				['compositionupdate', LiveCodeEvents._handleComposition],
-				['compositionend', LiveCodeEvents._handleComposition],
+				['compositionstart', LiveCodeEvents._handleComposition, true],
+				['compositionupdate', LiveCodeEvents._handleComposition, true],
+				['compositionend', LiveCodeEvents._handleComposition, true],
 				
-				['contextmenu', LiveCodeEvents._handleContextMenu],
+				['contextmenu', LiveCodeEvents._handleContextMenu, false],
 			];
 
 			var mapLength = mapping.length;
 			for (var i = 0; i < mapLength; i++) {
-				func(mapping[i][0], mapping[i][1]);
+				func(mapping[i][0], mapping[i][1], mapping[i][2]);
 			}
 		},
 
 		addEventListeners: function(pElement)
 		{
-			LiveCodeEvents._eventForEach(function(type, handler) {
-				pElement.addEventListener(type, handler, true);
+			// Create the hidden input element
```

Comment:
Hi @livecodeian ,
After one of these hidden inputs has been created, does it need destroying again once it is done with? Or is only one ever created and overwritten each time? Thinking about memory conservation.
Great work btw. Lots done here.
Sean.

Reply:
Hi @seaniepie, only one hidden input is created for the lifetime of the stack and is removed when the event listeners is removed.